lib/method_object.rb

```diff
@@ -1,53 +1,165 @@
+# frozen_string_literal: true
 require 'method_object/version'
 
+# See gemspec for description
 class MethodObject
+  class AmbigousMethodError < NameError; end
+
   class << self
     def attrs(*attributes)
-      define_attr_accessors(attributes)
-      define_setup_methods(attributes)
+      @attributes = attributes
+      Setup.call(attributes: attributes, subclass: self)
+    end
+
+    def call(**args)
+      new(args).call
+    end
+
+    attr_reader(:attributes)
+
+    private(:new)
+  end
+
+  def initialize(_); end
+
+  def call
+    raise NotImplementedError, 'Define the call method'
+  end
+
+  def method_missing(name, *args, &block)
+    candidates = candidates_for_method_missing(name)
+    case candidates.length
+    when 0
+      super
+    when 1
+      define_and_call_new_method(candidates.first)
+    else
+      handle_ambiguous_missing_method(candidates, name)
+    end
+  end
+
+  def respond_to_missing?(name)
+    candidates_for_method_missing(name).length == 1
```

Comment:
Rubocop told me to make this method but I didn't add a test for it. Comment out and TDD it back into existence.